include/deal.II/base/array_view.h

```diff
@@ -0,0 +1,219 @@
+   * the <em>view object</em>. It may however return a reference to
+   * a non-@p const memory location depending on whether the template
+   * type of the class is @p const or not.
+   */
+  value_type &operator[] (const std::size_t i) const;
+
+private:
+  /**
+   * A pointer to the first element of the range of locations in
+   * memory that this object represents.
+   */
+  value_type  *const starting_element;
+
+  /**
+   * The length of the array this object represents.
+   */
+  const std::size_t  n_elements;
+};
+
+
+
+//---------------------------------------------------------------------------
+
+
+template <typename ElementType>
+inline
+ArrayView<ElementType>::ArrayView(value_type        *starting_element,
+                                  const std::size_t  n_elements)
+  :
+  starting_element (starting_element),
+  n_elements(n_elements)
+{}
+
+
+
+template <typename ElementType>
+inline
+ArrayView<ElementType>::ArrayView(const ArrayView<typename boost::remove_cv<value_type>::type> &view)
+  :
+  starting_element (&view[0]),
+  n_elements(view.size())
+{}
+
+
+
+template <typename ElementType>
+inline
+std::size_t
+ArrayView<ElementType>::size() const
+{
+  return n_elements;
+}
+
+
+template <typename ElementType>
+inline
+typename ArrayView<ElementType>::value_type &
+ArrayView<ElementType>::operator[](const std::size_t i) const
+{
+  Assert (i<n_elements, ExcIndexRange(i, 0, n_elements));
+
+  return *(starting_element + i);
+}
+
+
+/**
+ * Create a view to an entire std::vector object. This is equivalent
+ * to initializing the current object with a pointer to the first
+ * element and the size of the given argument.
+ *
+ * This function is used for non-@p const references to objects of vector type.
+ * Such objects contain elements that can be written. Consequently, the return
+ * type of this function is a view to a set of writable objects.
+ *
+ * @param[in] vector The vector for which we want to have an array
+ *   view object. The array view corresponds to the <em>entire</em>
+ *   vector.
+ *
+ * @relates ArrayView
+ */
+template <typename ElementType>
+inline
+ArrayView<ElementType>
+make_array_view (std::vector<ElementType> &vector)
+{
+  return ArrayView<ElementType> (&vector[0], vector.size());
+}
+
+/**
+ * Create a view to an entire std::vector object. This is equivalent
+ * to initializing an ArrayView object with a pointer to the first
+ * element and the size of the given argument.
+ *
+ * This function is used for @p const references to objects of vector type
+ * because they contain immutable elements. Consequently, the return type
+ * of this function is a view to a set of @p const objects.
+ *
+ * @param[in] vector The vector for which we want to have an array
+ *   view object. The array view corresponds to the <em>entire</em>
+ *   vector.
+ *
+ * @relates ArrayView
+ */
+template <typename ElementType>
+inline
+ArrayView<const ElementType>
+make_array_view (const std::vector<ElementType> &vector)
```

Comment:
can we get a test for this please? Especially const vs non-const in argument and ArrayView.